proto/cosmos_proto/cosmos.proto

```diff
@@ -1,18 +1,88 @@
 }
 
 extend google.protobuf.FieldOptions {
-    // accepts_interface is used to annote fields that accept interfaces
-    string accepts_interface = 93001;
-    // scalar is used to define scalar types
-    string scalar = 93002;
+
+    // accepts_interface is used to annotate that a google.protobuf.Any
+    // field accepts messages that implement the specified interface.
+    // Interfaces should be declared using a declare_interface file option.
+    optional string accepts_interface = 93001;
+
+    // scalar indicates the
+    optional string scalar = 93002;
+}
+
+extend google.protobuf.FileOptions {
+
+    // declare_interface declares an interface type to be used with
+    // accepts_interface and implements_interface. Interface names are
+    // expected to follow the following convention such that their declaration
+    // can be discovered by tools: for a given interface type a.b.C, it is
+    // expected that the declaration will be found in a protobuf file named
+    // a/b/interfaces.proto in the file descriptor set.
+    repeated InterfaceDescriptor declare_interface = 793021;
+
+    // declare_scalar declares a scalar type to be used with
+    // the scalar field option. Scalar names are
+    // expected to follow the following convention such that their declaration
+    // can be discovered by tools: for a given scalar type a.b.C, it is
+    // expected that the declaration will be found in a protobuf file named
+    // a/b/scalars.proto in the file descriptor set.
+    repeated ScalarDescriptor declare_scalar = 793022;
+}
+
+// InterfaceDescriptor describes an interface type to be used with
+// accepts_interface and implements_interface and declared by declare_interface.
+message InterfaceDescriptor {
+
+    // name is the name of the interface. It should be a short-name (without
+    // a period) such that the fully qualified name of the interface will be
+    // package.name, ex. for the package a.b and interface named C, the
+    // fully-qualified name will be a.b.C.
+    required string name = 1;
+
+    // description is a human-readable description of the interface and its
+    // purpose.
+    optional string description = 2;
+}
+
+// ScalarDescriptor describes an scalar type to be used with
+// the scalar field option and declared by declare_scalar.
+// Scalars extend simple protobuf built-in types with additional
+// syntax and semantics, for instance to represent big integers.
+// Scalars should ideally define an encoding such that there is only one
+// valid syntactical representation for a given semantic meaning,
+// i.e. the encoding should be deterministic.
+message ScalarDescriptor {
+
+    // name is the name of the scalar. It should be a short-name (without
+    // a period) such that the fully qualified name of the scalar will be
+    // package.name, ex. for the package a.b and interface named C, the
+    // fully-qualified name will be a.b.C.
+    required string name = 1;
```

Comment:
we're going with the solution with a separate file that will map `a.b.C` to the actual language-specific scalar/interface, right?

Reply:
For scalars actually, the idea is that we're going to expect a go type `C` in the go package that the proto package maps to. ex. if `a.b` maps to `foo.bar/a/b` then there should be a go type `foo.bar/a/b.C`. For interfaces this model doesn't care because we're trying to deprecate the definition of go interfaces on proto types